### PR TITLE
TP-653 Edit additional codes

### PR DESCRIPTION
**Pull request by DIT-CJWaszczuk**
## Sumary 
Allows a user with permissions to edit additional codes
### Details page 
image

### Edit page 
image

### Confirmation page
image

### CODE REVIEW

**Comment by andyhd** [comment]
This build error is fixed in the PR I just merged, so you should be able to just rebase and it will work

**Review by simonwo — APPROVED** [review]

File: additional_codes/models.py [hunk 1]
```diff
@@ -83,6 +83,9 @@ class AdditionalCode(TrackedModel, ValidityMixin):
         business_rules.ACN17,
     )
 
+    def __str__(self):
+        return self.type.sid + self.code
```
Comment (andyhd) [inline additional_codes/models.py:87]: It's better to use `f"{self.type.sid}{self.code}"`

File: additional_codes/views.py [hunk 2]
```diff
@@ -30,14 +39,31 @@ class AdditionalCodeViewSet(viewsets.ReadOnlyModelViewSet):
     ]
 
 
-class AdditionalCodeList(TamatoListView):
+class AdditionalCodeTypeViewSet(viewsets.ReadOnlyModelViewSet):
+    """API endpoint that allows additional code types to be viewed."""
+
+    queryset = AdditionalCodeType.objects.latest_approved()
+    serializer_class = AdditionalCodeTypeSerializer
+    permission_classes = [permissions.IsAuthenticated]
```
Comment (andyhd) [inline additional_codes/views.py:47]: I'm not sure you need to add this - this is the API viewset and not relevant to this ticket. Also, we probably don't want the queryset to be `latest_approved`, but I'm not sure.

Comment (DIT-CJWaszczuk) [reply additional_codes/views.py:47]: This already existed, I just moved it in the file so it went API classes -> UI classes. Was previously bunched in between

Comment (andyhd) [reply additional_codes/views.py:47]: Never mind, I see this is just moved

**Review by andyhd — APPROVED** [review]

File: additional_codes/views.py [hunk 3]
```diff
@@ -48,19 +74,30 @@ class AdditionalCodeList(TamatoListView):
...
+    def get_object(self, queryset: Optional[models.QuerySet] = None) -> models.Model:
+        obj = super().get_object(queryset)
+
+        if self.request.method == "POST":
+            obj = obj.new_draft(
+                WorkBasket.current(self.request),
+                save=False,
+            )
+
+        return obj
```
Comment (andyhd) [inline additional_codes/views.py:99]: Looks like we can pull this out into `DraftUpdateView`

**Review by andyhd — APPROVED** [review]

File: common/jinja2/common/edit.jinja [hunk 4]
```diff
@@ -1,34 +1,18 @@
...
     "items": [
       {"text": "Home", "href": url("index")},
-      {"text": object._meta.verbose_name_plural|capitalize, "href": object.get_url("list")},
-      {"text": object._meta.verbose_name|capitalize ~ " " ~ object|string, "href": object.get_url()},
+      {"text": "Find and edit " ~ object._meta.verbose_name_plural, "href": object.get_url("list")},
+      {"text": object._meta.verbose_name|capitalize ~ ": " ~ object|string, "href": object.get_url()},
       {"text": page_title}
     ]
   }) }}
 {% endblock %}
```
Comment (andyhd) [inline common/jinja2/common/edit.jinja:14]: This block is the same as the detail page and the confirm-update page, so maybe they can all extend a common base template. Not at all urgent, though.

File: additional_codes/views.py (hunk 3 above)
Comment (andyhd) [reply additional_codes/views.py:99]: I was wrong 😁

File: common/jinja2/common/edit.jinja (hunk 4 above)
Comment (DIT-CJWaszczuk) [reply common/jinja2/common/edit.jinja:14]: I have a feeling the confirm may change when we have create pages, so will wait till it potentially gets generalised then.

**Pull request merged**